### PR TITLE
Complete type mapping in TypeScript backend

### PR DESCRIPTION
**Pull request by ssssam**
See https://github.com/finos/morphir-elm/issues/503#issuecomment-922860280 for the full mapping.

  * Generated TypeScript code works when imported explicitly, but you cannot import Morphir.IR as as a single module. This will be resolved in a later PR.
  * The `Char` type a placeholder - we have no built-in TypeScript type that represents a char. Using `string` for now as a workaround.
  * Code that verifies/exercises the generated types using TypeScript is done in https://github.com/finos/morphir-elm/pull/532.
  * The `Dict<key,value>` type becomes `Array<key,value>`, which works but is not the best type to use. We should use [Map](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map).

# Terms

> THIS SOFTWARE IS CONTRIBUTED SUBJECT TO THE TERMS OF THE TERMS OF THE CCLA DATED 2017-11-07 WITH FINOS/LINUX FOUNDATION (FORMERLY THE SYMPHONY SOFTWARE FOUNDATION CCLA).
>
> THIS SOFTWARE IS LICENSED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE AND ANY WARRANTY OF NON-INFRINGEMENT, ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE. THIS SOFTWARE MAY BE REDISTRIBUTED TO OTHERS ONLY BY EFFECTIVELY USING THIS OR ANOTHER EQUIVALENT DISCLAIMER IN ADDITION TO ANY OTHER REQUIRED LICENSE TERMS.

### CODE REVIEW

File: src/Morphir/TypeScript/Backend.elm [hunk 1]
```diff
@@ -81,93 +83,155 @@ mapModuleDefinition opt distribution currentPackagePath currentModulePath access
             , typeDefs = typeDefs
             }
     in
-    [moduleUnit]
+    [ moduleUnit ]
+
+
+{-| Map a Morphir Constructor (A tuple of Name and Constructor Args) to a Typescript AST Interface
```
Comment (ssssam) [inline src/Morphir/TypeScript/Backend.elm:89]: I suggest rewording this:

> Map one Constructor from a Morphir custom type to a TypeScript interface.

File: src/Morphir/TypeScript/Backend.elm [hunk 2]
```diff
@@ -81,93 +83,155 @@ mapModuleDefinition opt distribution currentPackagePath currentModulePath access
...
+                Private ->
+                    TS.Private
+    in
     case typeDef.value.value of
-        Type.TypeAliasDefinition typeArgs typeExp ->
+        Type.TypeAliasDefinition variables typeExp ->
             [ TS.TypeAlias
-                (name |> Name.toTitleCase)
-                (typeExp |> mapTypeExp)
+                { name = (name |> Name.toTitleCase)
```
Comment (Cynical-Optimist) [inline src/Morphir/TypeScript/Backend.elm:135]: The parentheses around `name |> Name.toTitleCase` are unnecessary, and should have been removed by autoformatting. This doesn't match up with the current head of dwinship/Typescript-Backend.

There's a few other autoformatting changes between this branch and the unsquashed branch.

(I think I may have confused things during a rebase in the last couple of days. I think I added the autoformatting changes, but edited the changes into one of the existing commits, rather than giving them their own commit. That might have contributed to this.)

File: src/Morphir/TypeScript/Backend.elm (hunk 1 above)
Comment (Cynical-Optimist) [reply src/Morphir/TypeScript/Backend.elm:89]: That suggests that the input is a Morphir custom type, and the output is a Typescript interface. Strictly speaking, the input is only part of a Morphir custom type.

At the time I also found it really helpful to clarify that a constructor consists of a tuple of (Name, List[args]) because that isn't clearly documented anywhere else in our code base. That sort of structural information was what confused me the most when I first tried to understand the code Attila had written.

I suppose it's not accurate any more though (now there's also the variables to consider). So you might be right that that needs to change.

File: src/Morphir/TypeScript/Backend.elm (hunk 2 above)
Comment (Cynical-Optimist) [reply src/Morphir/TypeScript/Backend.elm]: Huh, that's weird. I was wrong about dwinship/Typescript-Backend having all the autoformat fixes. It had some, but not all of them.

I've rebased and updated both of them to have all the autoformatting now, and the Typescript Backend files are identical on both branches now.

**Review by AttilaMihaly — APPROVED** [review]

File: src/Morphir/TypeScript/Backend.elm (hunk 2 above)
Comment (ssssam) [reply src/Morphir/TypeScript/Backend.elm]: Confirmed that `elm-format` doesn't make any changes on commit ccfc7b8bb9b890132c5920c43554db5b5a801953. I probably introduced this problem when I squashed and rebased your work, thanks for fixing.

File: src/Morphir/TypeScript/Backend.elm (hunk 1 above)
Comment (ssssam) [reply src/Morphir/TypeScript/Backend.elm:89]: > Strictly speaking, the input is only part of a Morphir custom type.

True.

> At the time I also found it really helpful to clarify that a constructor consists of a tuple of (Name, List[args]) because that isn't clearly documented anywhere else in our code base. That sort of structural information was what confused me the most when I first tried to understand the code Attila had written.

It's the job of the type system and variable names to provide this info. I agree the function declaration is hard to read, but can we fix by adding some meaningful type aliases rather than a comment?

Comment (Cynical-Optimist) [reply src/Morphir/TypeScript/Backend.elm:89]: That sounds good

**Comment by ssssam** [comment]
This is superceded by https://github.com/finos/morphir-elm/pull/533
